DSHOT/Bitbang - Hold the line at idle for 1 extra bit period.

hydra opened the pull request:
Fix DSHOT/Bitbang not holding the line at idle level for the full bit-period.

After the last transfer of the last DMA word which sets the GPIO BSRR register to IDLE an IRQ is raised for DMA transfer complete.  The handler then transitions the output to INPUT.  On fast MCUs this transition will occur within the last bit-period of the DSHOT frame, while the ESC is apparently sampling the signal.

This PR holds the line at idle for one complete bit period longer after the last dma transfer for the last bit.  which results in 66% of a DSHOT bit.  66% not 100% as the DMA transfer complete interrupt still occurs after the last transfer.  We could perhaps reduce this to 1 extra transfer as I originally remarked, instead of 3, to get the extra 33% that we need to complete the HOLD state of the last bit.

Before:
![DS1Z_QuickPrint34](https://user-images.githubusercontent.com/57075/158017608-d36638e9-855d-4404-9116-fd65553c6f77.png)

After:
![DS1Z_QuickPrint33](https://user-images.githubusercontent.com/57075/158017618-4cfd4c61-50e1-4401-9c08-6fa8b4f6afa1.png)

Details:
![DS1Z_QuickPrint36-annotated-2](https://user-images.githubusercontent.com/57075/158017620-10cba92a-6610-4578-bad7-b916096aa869.png)

For discussion, see here:

https://github.com/betaflight/betaflight/issues/11155#issuecomment-1065867093

hydra commented:
Binaries for testing:
[betaflight_4.3.0_SPRACINGH7EXTREME_25023bb3a2.zip](https://github.com/betaflight/betaflight/files/8237516/betaflight_4.3.0_SPRACINGH7EXTREME_25023bb3a2.zip)
[betaflight_4.3.0_SPRACINGH7NANO_25023bb3a2.zip](https://github.com/betaflight/betaflight/files/8237604/betaflight_4.3.0_SPRACINGH7NANO_25023bb3a2.zip)
[betaflight_4.3.0_SPRACINGH7ZERO_25023bb3a2.zip](https://github.com/betaflight/betaflight/files/8237605/betaflight_4.3.0_SPRACINGH7ZERO_25023bb3a2.zip)
[betaflight_4.3.0_STM32H743_25023bb3a2.zip](https://github.com/betaflight/betaflight/files/8237517/betaflight_4.3.0_STM32H743_25023bb3a2.zip)
[betaflight_4.3.0_MATEKH743_25023bb3a2.zip](https://github.com/betaflight/betaflight/files/8237515/betaflight_4.3.0_MATEKH743_25023bb3a2.zip)

hydra commented:
@jflyper As the original author of the BitBang code can you review please?

@4712 @alkamotors @sskaug @joelucid Can you confirm the *exact* times/offsets that your ESC firmware samples the signal at?

hydra commented:
DSHOT ESC authors, please see this comment:

https://github.com/betaflight/betaflight/issues/11155#issuecomment-1065881112

mathiasvr commented:
In BLHeli_S (and bluejay) a timer gated by the signal is used to measure the pulse widths/durations. I'm not exactly sure what you mean by the time/offset the signal is sampled?

However, the idle period at the end of the frame should be important because the signal must not transition for around 1.5x period (5.2us for dshot300), otherwise an extra pulse/bit might become part of the frame and it will be rejected for not being exactly 16 bits.

MoppelMat commented:
> Binaries for testing:
> [betaflight_4.3.0_SPRACINGH7EXTREME_25023bb3a2.zip](https://github.com/betaflight/betaflight/files/8237516/betaflight_4.3.0_SPRACINGH7EXTREME_25023bb3a2.zip)
> [betaflight_4.3.0_SPRACINGH7NANO_25023bb3a2.zip](https://github.com/betaflight/betaflight/files/8237604/betaflight_4.3.0_SPRACINGH7NANO_25023bb3a2.zip)
> [betaflight_4.3.0_SPRACINGH7ZERO_25023bb3a2.zip](https://github.com/betaflight/betaflight/files/8237605/betaflight_4.3.0_SPRACINGH7ZERO_25023bb3a2.zip)
> [betaflight_4.3.0_STM32H743_25023bb3a2.zip](https://github.com/betaflight/betaflight/files/8237517/betaflight_4.3.0_STM32H743_25023bb3a2.zip)
> [betaflight_4.3.0_MATEKH743_25023bb3a2.zip](https://github.com/betaflight/betaflight/files/8237515/betaflight_4.3.0_MATEKH743_25023bb3a2.zip)

to try to test I would need a binary for the Target MAMBAH743, please.

daleckystepan commented:
MAMBAF722_I2C dropped error rate from 20% to under 1% as on screenshot.

![IMG_0762](https://user-images.githubusercontent.com/36531759/158026604-4ff98c53-07bf-48ab-8759-99f0837897b1.png)

hydra commented:
> In BLHeli_S (and bluejay) a timer gated by the signal is used to measure the pulse widths/durations.

@mathiasvr yeah, there's two ways of doing it that I can think of:
1. using capture compare channels to measure pulse widths and recording the capture compare values at timer overflow into a DMA buffer.
2. using timer triggered on the first edge which then triggers a second timer after a delay from the first timer (i.e. at 50% of the period), update events on the second timer triggers a DMA transfer from the GPIO IDR register into a DMA buffer.

sounds like you're taking the first approach where you measure the pulse widths.  in which case the measured widths should be acceptable as a 0 if they are less than 37.5%, acceptable as a 1 if they are between 37.5% and 75%, with appropriate margin for error.  in this case the 33% and 66% out-of-spec widths fall into these categories.

@daleckystepan Exactly what ESC were you using and what firmware version does it have?

With this PR the DShot output signal is held for the correct amount of time, we don't want to hold the signal any longer because of the increase memory reads and DMA transfers and latency of transitioning the output to an input.  

IMHO, given that the output signal is now correct that the onus is on ESC software authors to ensure their firmware doesn't allow any glitches that occur after the frame to affect the readings.

Scenario 2 above achieves this by design. For 1 using timer gating/blanking to avoid sampling *anything* after the last bit.

It would be good to get confirmation from ESC authors that their firmware can handle these post-frame glitches and perhaps update the BF DSHOT wiki page with a list of firmware that CAN handle it so users are informed of compliance.

hydra commented:
> to try to test I would need a binary for the Target MAMBAH743, please. never mind. already flashed the generic stm32 bin

@MoppelMat yes, that's why I built it 😄   Please report back with your testing results.

daleckystepan commented:
@hydra It is this stack: https://www.unmannedtechshop.co.uk/product/mamba-basic-f722-mini-mk3-40a-flight-stack-20x20/ with BlHeli 32.83 BETA

hydra commented:
@haslinghuis Can this be merged into a new 4.2 maintenance branch too?  I'm sure those wanting a stable firmware would appreciate this fix while 4.3.x is being finalized.

mathiasvr commented:
For BLHeli_S it's simple to allow the post frame glitch since it's literally just removing a check that's made to not allow this. However, is that a good idea? I think it's there to require a stable signal instead of a glitchy one. IMO the FC firmware should be made to switch mode without glitch, instead of the ESC firmware having to handle this?

It's already working around other things maybe because of poor dshot spec and betaflight setting the standard. This can generally be quite complicated on BLHeli_S ESCs e.g. it doesn't even have DMA.

hydra commented:
> For BLHeli_S it's simple to allow the post frame glitch since it's literally just removing a check that's made to not allow this. However, is that a good idea? I think it's there to require a stable signal instead of a glitchy one. IMO the FC firmware should be made to switch mode without glitch, instead of the ESC firmware having to handle this?
> 
> It's already working around other things maybe because of poor dshot spec and betaflight setting the standard. This can generally be quite complicated on BLHeli_S ESCs e.g. it doesn't even have DMA.

@mathiasvr I agree that it's best if the signal is clean, but if there's no way to provide a clean signal then there's no choice other than to ditch the standard or work around it.  I'm open to suggestions.  For the non-bitbang approach the solution on the H7 was to go from AF -> OUTPUT -> INPUT mode, but for bitbang it's already *in* output mode, *and* driving the line.  But that was only for H7 and as @daleckystepan commented above it happens with BitBang on F7 too...

Perhaps this is another nail in the coffin for DSHOT and we all need to move to S2M implemented by FETTEC which, apparently, has lower processing requirements on ESCs.  The DSHOT system, while better than analog, does appear to have many issues.

Perhaps we should update the DSHOT standard to acknowledge the fact that for bi-directional systems glitches can occur , and are allowed to occur, during the IO transition period? Where's the DSHOT standard document again? ...

mathiasvr commented:
Yes I wasn't aware that you could not change the input/output mode without pulling the signal in the wrong direction on these.

I have actually been drafting another protocol as well to support more telemetry and lower processing requirements in regards to some feature requests. I hadn't heard of S2M so maybe this is already a better alternative.

With this fix is it even necessary to change any ESC dshot handling? I don't mind updating it if possible, but it might be nice to keep it back compatible with earlier ESC/versions.

hydra commented:
> Yes I wasn't aware that you could not change the input/output mode without pulling the signal in the wrong direction on these.

Yeah, the PUPD and MODER registers take effect immediately.  It would be nice if you could do a single write to one register to change both PUPD and MODE bits, but no...

> I have actually been drafting another protocol as well to support more telemetry and lower processing requirements in regards to some feature requests. I hadn't heard of S2M so maybe this is already a better alternative.

Check the FETTEC discord server where there is documentation available - https://discord.gg/SUDQKfFneD

> With this fix is it even necessary to change any ESC dshot handling? I don't mind updating it if possible, but it might be nice to keep it back compatible with earlier ESC/versions.

This change does not change any timings, it simply holds the line at idle state for 1 bit period longer, previously it was 'letting go' of the line too early.  DSHOT BitBang uses 33%/66%, DSHOT DPWM uses 37.5%/75%.  This has been the case for a long time now.  There are many FCs out there, the default may be to use BitBang on some FCs and DPWM on others too (!).

AlkaMotors commented:
@hydra sorry, late to the conversation.  What should I be looking for? I just tested a f7 with bi-dir bit bang on and off. Is this only with H7 I don't see that frame glitch to be an issue. The dma buffer will be full and that glitch won't be captured. 33/66 or 37/75 should both be fine.

sskaug commented:
BLHeli_32 behaves very similar to BLHeli_S. Except that for '32 of course the timer pulse width captures are done with DMA and not interrupts. The specs for the pulsewidths are quite wide, similar to 'S, I too will have to look up the exact criteria, but that is not the core of the issue here.
The reason why there is an issue here, is that the Dshot frame sometimes is read by the ESC as 13bits (DMA transfers for '32). And for data integrity robustness, BLHeli not only looks for correct XOR bits (which is a really weak error detection), but also requires the frame to have 12bits and no more and no less.
@hydra, I don't understand why you say you change PUPD and MODER? Isn't PUPD set for pullup all the time (it is in the ESC)?

hydra commented:
@sskaug Why is the frame sometimes read as 13 bits *after* this PR is applied?    Also, do you mean 17?  There are 12 data bits and 4 crc bits, 16 bit frame.

When using DMA the timer that triggers the DMA read should be started on the first bit, and then overflow 16, 48 times, once for each bit or once for each state.  If the DMA stream is configured with the number of transfers to be 16 or 48 it will stop at the end of transfer and no more bits will be in the DMA buffer.

Given that IO transition glitches are known to occur, it doesn't seem possible to exclude the possibility of an edge in a would-be-17th-bit period.  Lengthening the frame on the FC which has costs we don't want, as previously mentioned, and tightens the required timing on the FC between the last DMA transfer and the IO transition, which we also don't want.

With regards to the PUPD I was wondering if perhaps a sequence of changes to MODER and PUPD around the time of the IO transition might help, but I can't see how or where this might be done as the PUPD is set based on the configuration of the output pin and never changed.  This was possible in DPWM, an extra transition on the MODER was done, from AF (timer output level)->OUTPUT (fixed idle level)->INPUT.

`dshotBitbangDevInit` defines the `bbPuPdMode` like this:

```
        uint8_t output = motorConfig->motorPwmInversion ?  timerHardware->output ^ TIMER_OUTPUT_INVERTED : timerHardware->output;
        bbPuPdMode = (output & TIMER_OUTPUT_INVERTED) ? BB_GPIO_PULLDOWN : BB_GPIO_PULLUP;
```

The value computed for `bbPuPdMode` is later used in `bbGpioSetup` to set the PUPD register, I couldn't find any places where the code changed the values of the PUPD register, usually done via `IOConfigGPIO`.

For good measure I verified the PUPD and MODE registers on an H7EXTREME (PA0-PA3), there is a transition on the MODER/MODE0-3 from 0x0 (input) to 0x1 (output) and PUPD/PUPD0-3 remains at 0x1 (Pull-Up)

![image](https://user-images.githubusercontent.com/57075/158171687-c2acd353-5463-402f-8be3-c8651039eeb6.png)
![image](https://user-images.githubusercontent.com/57075/158171801-400f1b25-99d0-4f5c-9a9a-d9cf393e1294.png)

![image](https://user-images.githubusercontent.com/57075/158172275-1992b2eb-8664-4dda-8ba5-f2d1a43b09b8.png)
![image](https://user-images.githubusercontent.com/57075/158172226-5949cf8f-d35f-4883-a929-467c03cad758.png)

So yes, it is set up for PUPD all the time, so probably nothing we can do to remove the transition glitch for BitBang, and thus my thinking is that the ESC just needs to handle it.  At least now the transition glitch does not occur in the last bit period with this PR.

hydra commented:
> @hydra sorry, late to the conversation. What should I be looking for? I just tested a f7 with bi-dir bit bang on and off. Is this only with H7 I don't see that frame glitch to be an issue. The dma buffer will be full and that glitch won't be captured. 33/66 or 37/75 should both be fine.

Sounds perfect, as long as you don't sample anything after the 16th bit you will be fine.  No, not only with H7 as @daleckystepan noted this happens on an F722 as well.

Thanks for responding and confirmation that AK32 is Ok as-is.

daleckystepan commented:
@hydra I am not 100 % confident that it is the same issue on F7 but it seems to be as applying this PR helped.

hydra commented:
> @hydra I am not 100 % confident that it is the same issue on F7 but it seems to be as applying this PR helped.

Are you able to check the output from your F722 on scope, before and after this PR?  Either way this PR still changes the output signal so the line is held appropriately...

daleckystepan commented:
Unfortunately not. The improvement I have observed here was tested by my friend in different country.

 I have a scope so I can check it on my another F7 board. But I am not sure if it make any sense as I have no issues on that board.

sskaug commented:
@hydra. Yes, sorry, of course I meant 17bits and not 13. The "CRC" is not a CRC. It is only an XOR, and hence a very weak error detection, so in BLHeli we also use a criteria of correct frame length. Which of course is easy to remove, but not on the entire installed base of ESCs.
If at all possible, I strongly recommend that the FC issues a clean, glitch free signal. Have you tried things like filling the "17th bit" memory with "000" (instead of "1X0")? And maybe also actually also transmit that bit?

mathiasvr commented:
Since this issue was discovered because of the blheli check I think it's doing a good job of not letting discrepancies like this go unnoticed and instead be documented. It really would be nice to handle this where it originates instead of having a "do this because betaflight does that" comment in the ESC firmware which is backwards.
I don't exactly understand the problem with this workaround in betaflight but is it so bad that it justifies making the ESC firmware more likely to accept an actually unstable signal?
Not sure how to tell since issues with e.g. dshot signal interference doesn't seem to be very common but has come up before. If AM32 has not had any issues with this it might not be a problem. Would be great to have some experiments with signal corruption to have a measure of this. I think the xor check might be "good enough" in the sense that you let more valid packets through than invalid, however since this could still be considered dangerous in an extreme situation I understand if some would not consider optimizing betaflight more important.

Without more information it seems almost easiest for me to add a new setting of "signal corruption protection" low/high, which seems really strange, but in my case I don't mind letting the users decide since some uses it for other custom projects and it makes it possible for them to experiment with this which I don't have the time to do myself.

hydra commented:
> Unfortunately not. The improvement I have observed here was tested by my friend in different country.
> 
> I have a scope so I can check it on my another F7 board. But I am not sure if it make any sense as I have no issues on that board.

I would be good to inspect the output on of ALL motor outputs on your other F7 board on a scope using DShot bitbang and also giving details of the a) which exact IO pins were inspected and b) details of the CPU and it's revision and date-code.

daleckystepan commented:
@hydra Here is the requested info from my friend.

![IMG_3225](https://user-images.githubusercontent.com/36531759/158377966-c234d050-16d6-4af4-b78e-c64928340cca.jpeg)


[BTFL_cli_20220315_123200.txt](https://github.com/betaflight/betaflight/files/8253298/BTFL_cli_20220315_123200.txt)

AlkaMotors commented:
I just check the outputs on the H7 I have and also the F7. dshot bitbang is ON. I am not seeing any major glitches like what is show above any channel with dshot 300 or 600. Channel 2 ( Iflight H7-twing) has a slight dip in voltage after the signal but nothing that would trigger another edge to be detected. Not that it would matter anyway. I probably don't do things the right way but am32 records 32 switching edges. After that, the dma interrupt fires and the signal is checked to ensure the entire transfer is within tolerance for a dshot signal then proceeds with xor check. Anything that happens after that 32nd line change is not going to be recorded in the stream.

sskaug commented:
I'm not sure I was perfectly clear here. The proposal is not to actually necessarily send one additional bit. The request was that maybe adding one more memory location (with 000) (while actually sending 16 bits) could fix the bug in the combination of BF implementation and STM hardware that causes such a glitch on the signal output. Maybe this works, or maybe it is actually necessary to also send the 17th bit.
In BLHeli_32 code, we have done similar tweaks relating to MCU hardware implementations to avoid glitches on erpm return data.

sskaug commented:
I haven't followed this whole dialog from the beginning. Is it so that this is only an issue for the H7 MCUs?
To me the glitch looks like it happens because a new bit is "started" before the DMA is terminated. So the thought is to make sure this new bit (3 DMA transfers as I understand it) has data that will not change the output. In BLHeli_32 we now support a lot of different MCUs and have discovered on several occasions that there are various quirks in the digital hardware that requires modifications to how FW handles the HW. Unfortunately - because it does complicate code.

hydra commented:
> I haven't followed this whole dialog from the beginning. Is it so that this is only an issue for the H7 MCUs?

No, apparently happens on F7 too, but I've not personally seen it on there but this PR has been shown to improve the CRC situation with F7's too.

> To me the glitch looks like it happens because a new bit is "started" before the DMA is terminated.

That's not what's happening, the circuits in the FC MCU is dragging the voltage down when IO transition occurs, it always happens, I'm not aware of a workaround, waiting longer (by using more DMA transfers) won't REMOVE the glitch, that would only delay it, and at the same time adds load to the FC and shortens the IO transition period, which we do not want.

> So the thought is to make sure this new bit (3 DMA transfers as I understand it) has data that will not change the output. 

This PR adds 3 transfers that do not change the output already, this holds the line idle for the last 33% of the 16th bit (fixes the bug) and gives you 66% of a new bit at IDLE level before the IO transition MAY occur, when the ISR eventually gets serviced, which is why you see solid lines and faint lines on my before/after scope captures as the ISR latency is not fixed.

hydra commented:
> am32 records 32 switching edges. After that, the dma interrupt fires and the signal is checked to ensure the entire transfer is within tolerance for a dshot signal then proceeds with xor check. Anything that happens after that 32nd line change is not going to be recorded in the stream.

@AlkaMotors Yes, then AM32 won't receive the glitch that occurs and is immune to it.  As long as the tolerance checks handle the 33%/66% instead of 37.5% and 75% then everything is fine with both DSHOT DPWM and DSHOT BitBang.

KarateBrot commented on src/main/drivers/dshot_bitbang_impl.h:
```diff
@@ -51,9 +51,11 @@
 
 #define MOTOR_DSHOT_FRAME_BITS             16
 
+
```
```suggestion
```

sskaug commented:
Actually I'd say that when a GPIO can not be changed from output to input without a glitch, this is so serious that it should be reported to STM. Have you considered that? Maybe they even have proposals on how to avoid it?

hydra commented:
> Actually I'd say that when a GPIO can not be changed from output to input without a glitch, this is so serious that it should be reported to STM. Have you considered that? Maybe they even have proposals on how to avoid it?

When we first went though the process of working round the glitch for the DSHOT DPWM @joelucid and I did a *lot* of digging around and tried lots of things, no official contact though.  Might be worth a shot...

This doesn't however change the fact that it occurs now and that ESCs need to handle this situation NOW.  We may not get any workarounds from ST and we all should be prepared for that situation.

@sskaug Suggest you start prepping an update for BlHeli.

Quick-Flash commented:
@hydra this does not fix the issue with the mambah743, it is still only giving me 2 out of 4 motors to spin :(

justinflipflops commented:
@Quick-Flash  I reported that as well on one of the issues I thought this was related too. I didn't get a response there. What ESC do you have? I. The issues threads others claimed to not have the problem, but additionally had different ESCs than others with the issue. Seems in most cases it's a stack with a diatone mamba ESC, generally high pwm variants. I only have one other ESC tucked in a build. I'm tempted to pull it out and test.

Quick-Flash commented:
> @Quick-Flash  I reported that as well on one of the issues I thought this was related too. I didn't get a response there. What ESC do you have? I. The issues threads others claimed to not have the problem, but additionally had different ESCs than others with the issue. Seems in most cases it's a stack with a diatone mamba ESC, generally high pwm variants. I only have one other ESC tucked in a build. I'm tempted to pull it out and test. 

I've got the pyro f3 45a esc

hydra commented:
> I've got the pyro f3 45a esc

@Quick-Flash Do you have an oscilloscope and can you give us the signal traces with/without the ESC connected?

Quick-Flash commented:
> @Quick-Flash Do you have an oscilloscope and can you give us the signal traces with/without the ESC connected?

No. I've got a logic analyzer.

hydra commented:
> No. I've got a logic analyzer, but I could perhaps use one of the schools. How do you typically probe the outputs to an esc while it is connected?

@Quick-Flash That would be extremely helpful.  Just prod the scope probe to a motor signal pad, the backs of the connectors usually have large enough pads to get the probe on. just don't slip with the scope probe and short one to VBAT voltage. 😲   Use a 'delay' trigger or 'nth edge' trigger set to 16th falling edge on the scope.

hydra commented:
@sskaug Do you have an update for us, can you do a new test firmware for BlHeli that ignores the glitch?

sskaug commented:
We can do testcode. Which BLHeli_32 target do you want testcode for?

howels commented:
FYI when using BlueJay on my JHEH743 this change which disables a check allows bidir shot to work on my board with 4.3.0 RC3 https://github.com/mathiasvr/bluejay/commit/daf6c11a24a23dde9dae6bc3f4ff711b9c8074f1

haslinghuis commented:
@howels also check https://github.com/betaflight/betaflight/issues/11155#issuecomment-1079903473

howels commented:
> @howels also check [#11155 (comment)](https://github.com/betaflight/betaflight/issues/11155#issuecomment-1079903473)

Yup. Caught that, good to see our understanding of the H7 increasing :+1: 

Looks like this can be solved several ways, hopefully we can preserve the integrity of dshot but at least there's a workaround for those wanting to fly 4.3 in the meantime.

joelucid commented:
@hydra - sorry for being late to the party - have been incredibly busy. I remember when implementing Bidirectional DSHOT (both the timer and the bitbang (together with @jflyper) variants) I saw these types of issues all the time. Each processor behaved a little differently. But in the end there always was a way to avoid the glitches by massaging registers.

I can't believe this is not fixable. Does the code directly write the registers or does it use the STM wrappers to do it? Often they do extra stuff which causes the problems. 

I also seem to recall that some boards had specific issues - like a hardware pull-down on the board. It's interesting that Alka doesn't see the issue. This could point to such an issue with the board you're testing?

Modifying ESC firmwares to work around these glitches should be the absolute last option to consider IMO.

joelucid commented:
Looking at https://github.com/betaflight/betaflight/issues/11155 the glitches only happen for some DSHOT output lines, not all. Therefore I suspect that this can't be typical behavior for these processors. If a output to input transition would cause the line to be pushed down on a given chip I would expect that to be consistent between different output lines. Either the FC boards have external pull downs or there is a timing related issue which causes the glitch to appear on some but not all lines.

joelucid commented:
Might be interesting to test without an esc connected - just to exclude the possibility that the esc fw is pushing the line down. Not very likely though since at least the efm8's would not be quick enough. And a pull-down on a signal line contradicts the whole bootloader approach of blheli-s and thus jesc.

joelucid commented:
In any case if this is a regression in 4.3 it's unlikely it's esc related and likely that the glitches can be fixed.

haslinghuis commented:
@joelucid please also see this comment: https://github.com/betaflight/betaflight/issues/11155#issuecomment-1079903473

![image](https://user-images.githubusercontent.com/8344830/160296901-15d39c26-888d-48f9-a769-e7f522782b1f.png)

joelucid commented:
Yeah - if the glitch can't be avoided then lowering the speed of the output drive could at least attenuate it. Still a workaround not a fix.

haslinghuis commented:
If only observed on H7 we should revise (as you mentioned different timing between MCU's):

![image](https://user-images.githubusercontent.com/8344830/160297197-cb980d2f-4550-4099-ad17-47e878423b44.png)

haslinghuis commented:
@hydra would be interesting if you can see a difference on your scope lowering the GPIO speed or maybe fixing it better as @daleckystepan pointed out there are still some glitches.

haslinghuis commented:
@joelucid what is the required GPIO bandwidth anyway?

joelucid commented:
@haslinghuis depends on the dshot bitrate of course. A DSHOT bit is either a long or a short level up followed by a level down. So you need something like 3x the DSHOT bitrate. Bidirectional DSHOT uses a different protocol on the telemetry side with the express purpose of limiting the required bandwidth as it can't be supplied by the efm8 chips.

haslinghuis commented:
![image](https://user-images.githubusercontent.com/8344830/160300548-897363ea-6f1b-4725-8b74-986a557e6feb.png)

joelucid commented:
Yep - but it is a very expensive protocol. All these level transitions cost dearly. As some folks have discussed possible improvements above: I always thought changing the FC to ESC communication to use the same protocol as Bidirectional DSHOT on the telemetry side would greatly reduce the load on both FC and ESC. That would be a big improvement though it wouldn't change what's available on the feature side.

joelucid commented:
To make the improvements to flight performance I had envisioned when introducing rpm filtering we'd need another quantum leap ahead. Like we would need to know when the zero crossing occurs the moment it does. Then we could apply power to the esc in a closed loop targeting rpm and rpm' which would open up a whole new world in flight performance improvements.

But it's not easy to achieve. Running escs and FC on the same processor would be one way to do this. Or a more analog protocol requiring more than 2 lines to each esc.

Problem with rpm control is that it degrades towards low rpm when zero crossings can be many ms from switching phases. And that Bidirectional DSHOT adds a lot of additional latency.

Arne-W commented:
> Yeah - if the glitch can't be avoided then lowering the speed of the output drive could at least attenuate it. Still a workaround not a fix.

Not sure if would say its a workaround - enabling speeds of up to 200Mhz where only 2Mhz are needed does not sound useful to me. Looking at the traces @MoppelMat has done at different speed settings it seems to be somehow systematical. Perhaps you would need some external parts to drive the gpio at 100Mhz to 200Mhz safely?!

hydra commented:
> In any case if this is a regression in 4.3 it's unlikely it's esc related and likely that the glitches can be fixed.

@joelucid it's not a regression unless I'm mistaken. The bi-directional has apparently always done it.  Our original glitch workarounds was for the DPWM implementation, not BitBang.

Yes, it's not ESC related but see my comment above regarding responsibility of improving robustness for the entire DSHOT eco-system.

It would also be ideal if the authors of the DShot handbook could update it and make mention of IO glitches that *may* occur, now and on future hardware, at the time of IO transition.

daleckystepan commented:
I have asked my friend to test LOW and MEDIUM options on MAMBAF722_I2C. In both cases DHOST600 error rate dropped from 20 % to under 1 %. It is better but not fully fixed yet (there can be another issue of course). The same results as after applying https://github.com/betaflight/betaflight/pull/11469. So I would prefer #11494 over https://github.com/betaflight/betaflight/pull/11469 as it is cleaner way of doing it.

hydra commented:
> So I would prefer #11494 over #11469 as it is cleaner way of doing it.

@daleckystepan The original code has a bug in that the line is not held for the entire duration of the all the DSHOT bits, it stops to early.

IMHO both this and the other #11494 should be merged together to make it as robust as we possibly can, and then it's over to the ESC guys to work on things on their side, as appropriate.

Without this PR there is a possibility that a glitch can occur /during/ the last DSHOT bit.

sskaug commented:
I am now ready for generating testcode, please let me know which code(s) (exact name as given by BLHeliSuite32) that you would like to have testcode for.

sskaug commented:
Testcode posted here: https://github.com/bitdump/BLHeli/tree/master/BLHeli_32%20ARM/Betaflight%20bidir%20Dshot%20glitch%20issue

LordGG commented:
> @LordGG what's does BlHeliSuite32 ESC for your "ESCs Mamba F55_128K BL32" ESC please, so that @sskaug can make new test-code for you.

@sskaug Hi, my ESCs are MAMBA_F55_128K_4**1**10

Sorry if my question might sound dumb, but what is the purpose of this BLHeli32 testcode ? Is there also an issue with Mamba 128K ESCs ?

sskaug commented:
Your code now posted. There is no issue with Mamba ESCs.
The purpose of this code is for BLHeli_32 to ignore a glitch that Betaflight sometimes generate at the end of the Dshot frame.

hydra commented:
@LordGG Please use the BlHeli testcode provided by @sskaug (thanks!) in combination with the same firmware that originally showed the error and report back.

Arne-W commented:
Ok I found some time and did some testing! I have a Mamba stack consisting of Mamba MK4 H743 with BMI270 and a Mamba F55_128K F4201. @hydra I took your firmware (25023bb3a2) flashed it and tested against BLHeli 32.83 and it does not work as expected.
![notwork](https://user-images.githubusercontent.com/7621911/161761436-331f3ed8-134c-4175-8fa8-e2cb8a69332f.png)

Then updated to the new BLHeli testcode @sskaug posted and it works without any errors!
![works](https://user-images.githubusercontent.com/7621911/161761845-9e549910-bff0-418e-9e69-486ab3c7ba63.png)

Third test was the new BLHeli testcode against the original betaflight 4.3.0 RC3 code and also this setup works without any errors.
![works_rc3](https://user-images.githubusercontent.com/7621911/161763995-31f432df-3a85-47fd-803d-3a5308cff2d0.png)

All tests were done with motors connected and no issues. @sskaug Looks like the testcode can handle the glitches we have seen in the scope traces.

For my test setup the change of this PR does not make a difference. Nevertheless I completely agree with @hydra that this fix should be pulled in just to make sure the last bit is always long enough.

But I am not sure if its a good idea in general to work around those glitches in the ESC Firmware. We would never have seen this issue if BLHeli wasn't that restrictive. From my point of view it is a good thing to know when the signal quality starts degrading - but that is only my opinion.

hydra commented:
> All tests were done with motors connected and no issues. @sskaug Looks like the testcode can handle the glitches we have seen in the scope traces.
> 
> For my test setup the change of this PR does not make a difference. Nevertheless I completely agree with @hydra that this fix should be pulled in just to make sure the last bit is always long enough.

@Arne-W Your findings match my expectations of what would happen *exactly*.

> But I am not sure if its a good idea in general to work around those glitches in the ESC Firmware. We would never have seen this issue if BLHeli wasn't that restrictive. From my point of view it is a good thing to know when the signal quality starts degrading - but that is only my opinion.

Sort of... The signal was actually incorrect, and thus we noticed it, however now that the frame part of the signal IS correct the BlHeli still showed errors.  That to me implies an implementation issue in the BlHeli code which the new BlHeli test code appears to correct.  It would probably be possible on the ESC side to detect the glitch and report it, but I don't think there's any merit in that if the glitch occurs AFTER the frame which this PR corrects.

I'm happy with fixed BlHeli, fixed GPIO speed and fixed DSHOT frame.  🎉

haslinghuis commented:
If this fix needs 32.9 firmware we should update release / tuning notes. How about other firmware like AM32, Blheli_S, Bluejay, JLucid, JazzMaverick, APD, Fettec, (hope this covers all)

hydra commented:
> If this fix needs 32.9 firmware we should update release / tuning notes.

This PR does not need the new BlHeli firmware.  The new BL Heli firmware works with OLD BF and also this PR.  So everyone that doesn't upgrade to BF 4.3 can still benefit from working BlHeli32 fixed firmware and that should be on the BlHeli release notes, not the BF 4.3 release notes.
Once a new BlHeli is released it would however be advisable to make mention of it in the wiki but that is beyond the scope of this PR.

LordGG commented:
> Your code now posted. There is no issue with Mamba ESCs. The purpose of this code is for BLHeli_32 to ignore a glitch that Betaflight sometimes generate at the end of the Dshot frame.

> @LordGG Please use the BlHeli testcode provided by @sskaug (thanks!) in combination with the same firmware that originally showed the error and report back.

I got that stack (Mamba H743 BMI-270 + MAMBA_F55_128K_4110) out of my drawer and hooked some motors to it. I kept the same BF version (4.3 RC1).

- With latest Official BLHeli32 32.9 : still the same issue, 2 motors not working with Bidir DShot enabled
- With testcode MAMBA_F55_128K_**4110**_Multi_32_91.Hex, the miracle occured, all 4 motors working with no error ! 👍 

![image](https://user-images.githubusercontent.com/6628377/161858236-35f676c3-4d29-417a-a68d-9292b99a4d62.png)
Thanks guys !

haslinghuis commented:
@LordGG please also test with RC4 firmware.

LordGG commented:
> @LordGG please also test with RC4 firmware. Did you test with or without this PR or did 32.9 ESC firmware solve the problem using RC1?

Just tested with RC4, still working.

I tested without this PR. Official 32.9 did nothing. Testcode 32.91 provided here solved the problem with RC1 (and confirmed with RC4).

Is this PR included in RC4 ?

haslinghuis commented:
No it's still open but will be included next cycle (RC5) as there were some safety improvements needing to bring out RC4. Thanks for testing.

haslinghuis commented:
Having tested with RC1 also implies the ESC firmware itself was enough to solve your problem 😕

LordGG commented:
> Having tested with RC1 also implies the ESC firmware itself was enough to solve your problem 😕 🤔 And thanks for pointing out 32.9.1 as solution instead of 32.9

If it can help, I can flash back my ESCs to official 32.9 and test if this PR alone also solves the issue. (but I would need to be provided with the compiled firmware of the PR)

haslinghuis commented:
Building now...

LordGG commented:
@haslinghuis Something unexpected happened 🤔

While you were compiling this branch, I flashed back my ESCs to Official 32.9.
I wanted to check if i had 2 motors failing again and... no ! 4 motors were still working.
No doubt about the version of the ESCs.
Then I flashed RC1... and has expected, 2 motors were failing.
I flashed RC4 again to make sure : 4 motors working
So I flashed RC3 : same as RC1 --> 2 motors failing
I flashed RC4 again (just to be 100% sure dshot bidir was enabled) : still 4 motors working.
Finally I tested your compiled firmware : also 4 motors working.

To summerize : 

**- Testcode 32.91 + BF RC1 : 4 motors working, OK**
**- Testcode 32.91 + BF RC4 : 4 motors working, OK**
- Official 32.9 + BF RC1 : 2 motors working, NOT OK
- Official 32.9 + BF RC3 : 2 motors working, NOT OK
**- Official 32.9 + BF RC4 : 4 motors working, OK**
**- Official 32.9 + BF Custom PR Build : 4 motors working, OK**

Seems like something has changed between RC3 and RC4 that also fixed the issue. I hope it will make more sense to you, I was not expecting RC4 alone did the job.

haslinghuis commented:
@LordGG this one has solved it probably: https://github.com/betaflight/betaflight/pull/11494

LordGG commented:
> How about the difference in RPM errors reported between these options?

I only did short tests for a few seconds each time, but I had 0% error with every options.

MartinWa commented:
In case it helps:
I have a Diatone Roma F5 v2 with the MAMBAF722 FC. I have installed BLHeli32 v32.9.02 and Betaflight 4.3 RC4. Enabling Bidirectional Dshot on Dshot600 gives errors on two motors. Dshot300 works fine

haslinghuis commented:
@MartinWa please with this local firmware: https://dev.azure.com/Betaflight/Betaflight%20Nightlies/_build/results?buildId=6207&view=artifacts&pathAsName=false&type=publishedArtifacts

MartinWa commented:
> @MartinWa please with this local firmware: https://dev.azure.com/Betaflight/Betaflight%20Nightlies/_build/results?buildId=6207&view=artifacts&pathAsName=false&type=publishedArtifacts

Yes that build works:
![image](https://user-images.githubusercontent.com/497015/162286410-1d516a6f-1f7a-494e-9a22-8679c5d9b1b2.png)

haslinghuis commented:
Thanks for testing and reporting back. Still see some noise, but could be build related.

haslinghuis commented:
Build PR rebased on `master` for testing:

[betaflight_4.3.0_STM32H743_9f8e3f6e2.zip](https://github.com/betaflight/betaflight/files/8447274/betaflight_4.3.0_STM32H743_9f8e3f6e2.zip)
[betaflight_4.3.0_STM32G47X_9f8e3f6e2.zip](https://github.com/betaflight/betaflight/files/8447276/betaflight_4.3.0_STM32G47X_9f8e3f6e2.zip)
[betaflight_4.3.0_STM32F745_9f8e3f6e2.zip](https://github.com/betaflight/betaflight/files/8447277/betaflight_4.3.0_STM32F745_9f8e3f6e2.zip)
[betaflight_4.3.0_STM32F7X2_9f8e3f6e2.zip](https://github.com/betaflight/betaflight/files/8447278/betaflight_4.3.0_STM32F7X2_9f8e3f6e2.zip)
[betaflight_4.3.0_STM32F411_9f8e3f6e2.zip](https://github.com/betaflight/betaflight/files/8447281/betaflight_4.3.0_STM32F411_9f8e3f6e2.zip)
[betaflight_4.3.0_STM32F405_9f8e3f6e2.zip](https://github.com/betaflight/betaflight/files/8447282/betaflight_4.3.0_STM32F405_9f8e3f6e2.zip)

Arne-W commented:
@haslinghuis What is the difference between the BF4.3 RC4 and the nightly @MartinWa tested? I would expect them to show identical behavior!

haslinghuis commented:
@Arne-W The build provided is a rebase of this PR on master, The Azure build is based on the time it is build so doesn't include later additions.

Arne-W commented:
@haslinghuis You got me wrong - I meant the the 2 versions @MartinWa tested. Those were A: BF 4.3 RC4 and B: The Azure nighly you provided the link.
These version should be the same regarding the DSHOT behavior. They both have the GPIO-Speed fix #11494 but not the fix from this PR and therefore they should behave the same. But @MartinWa said the Azure nightly fixes the problem he saw with BF4.3 RC4.
That is what I am wondering about.

haslinghuis commented:
@Arne-W ICU. 🤔 Might have something to do with something like optimization in debug and release versions as it would confirm the difference?

MartinWa commented:
If needed I am happy to test more

haslinghuis commented:
@MartinWa please test also https://github.com/betaflight/betaflight/pull/11469#issuecomment-1092279746

Its based on this PR rebased on current master.

MartinWa commented:
With: Betaflight / STM32F7X2 (S7X2) 4.3.0 Apr  8 2022 / 00:47:17 (9f8e3f6e2) MSP API: 1.44
I get
![image](https://user-images.githubusercontent.com/497015/162475564-3ddbe99d-9023-42b9-bc1f-818abe3a1dfc.png)

haslinghuis commented:
Curious about testing with RC4 now again.

MartinWa commented:
> Curious about testing with RC4 now again.

Updated my post above

Arne-W commented:
@MartinWa Thanks for the testing so far :+1: . It would be very cool if you could do the same test with the BF4.3 RC3 as it uses other settings for the GPIO Speed. Moreover I have some questions regarding your results:

- The problems are always motor 1 and 2?
- Its always DSHOT600?
- Are the error values heavily varying or are those numbers more stable (the value behind E:)?
- Are the results reproducible over several (2-3) reboots?

SteveCEvans commented:
@MartinWa Why are we seeing anything other than 0.00% errors? I see no errors on an F722 and dev build 2663 (5th Apr). What ESC are you using?

haslinghuis commented:
@SteveCEvans debug versions work, but release build not. Using same code. As master is at same point as RC4.

MartinWa commented:
> @MartinWa Thanks for the testing so far 👍 . It would be very cool if you could do the same test with the BF4.3 RC3 as it uses other settings for the GPIO Speed. Moreover I have some questions regarding your results:
> 
> * The problems are always motor 1 and 2?
Yes

> * Its always DSHOT600?
Yes, DSHOT300 has 0% errors on all versions

> * Are the error values heavily varying or are those numbers more stable (the value behind E:)?
On very low RPM the errors are higher but otherwise they are quite similar

> * Are the results reproducible over several (2-3) reboots?
Yes, I have now tested rebooting several times on each version

MartinWa commented:
Did more testing:
These tests are run on a Diatone Roma F5 v2 
config: manufacturer_id: DIAT, board_name: MAMBAF722_I2C, version: ec268d11, date: 2021-07-26T13:10:45Z
board: manufacturer_id: DIAT, board_name: MAMBAF722_I2C

For all the versions I have done the following
1. Flash with full chip erase
1. Apply custom defaults
1. Calibrate accelerometer
1. Get version from cli
1. Enable bidirectional dshot
1. Agree to filter settings
1. Save and reboot
1. Enable motors and test running on low up to mid rage of thrust on all four motors

Here are the results:
Betaflight / STM32F7X2 (S7X2) 4.3.0 Feb  1 2022 / 20:14:54 (3267f0417) MSP API: 1.44 (RC3)
![motor](https://user-images.githubusercontent.com/497015/162517341-02952e39-e55b-4d73-a9bc-3b1e9291a666.png)

Betaflight / STM32F7X2 (S7X2) 4.3.0 Apr  6 2022 / 07:58:20 (2fd2e8eca) MSP API: 1.44 (RC4)
![motor](https://user-images.githubusercontent.com/497015/162517475-5af9d826-4a50-4afc-9af8-e0844e035c2a.png)

Betaflight / STM32F7X2 (S7X2) 4.3.0 Apr  8 2022 / 00:47:17 (9f8e3f6e2) MSP API: 1.44 (PR)
![motors](https://user-images.githubusercontent.com/497015/162517489-b81bc6fc-42bf-4993-8f15-4d93f7054c18.png)

haslinghuis commented:
Did a hex compare on master against a RC4 build and indeed there is a difference:

![image](https://user-images.githubusercontent.com/8344830/162536036-ef420999-f529-41a9-baf6-488fb93951ef.png)

Arne-W commented:
Looking at the test results I would say there is no significant difference between RC3 and RC4 - which is very good, because the errors don't seem to depend on the GPIO speed. I suspect the problem is due to electrical interference, as the errors disappear when switching to DSHOT300.
Interestingly, this PR seems to greatly improve the behavior. It somehow seems to reduce the electrical noise. I have no idea why, but maybe @hydra has one?

@MartinWa Thank you for all the testing :+1: !

hydra commented:
> I suspect the problem is due to electrical interference, as the errors disappear when switching to DSHOT300.

Do you have a scope? Would like to see the scope traces when the motors are spinning and you're getting CRC errors.

> Interestingly, this PR seems to greatly improve the behavior, It somehow seems to reduce the electrical noise.

If your DSHOT DMA callback handler is being serviced quickly and the IO transitions cause glitches then all that will happen is that the glitch is delayed, it will still be there if the GPIO speed change doesn't fix the glitches for you.  Depending on ESC firmware you may get better results.  What ESC and ESC firmware are you using?

> I have no idea why, but maybe @hydra has one?

I don't have a Diatone Roma F5 v2, can someone get one shipped to me?

howels commented:
FYI the diatone Roma V2 appears to have this stack with the F50 ESC option: https://www.diatone.us/products/mamba-f722s-flight-controller-stack?variant=39675134574679

MartinWa commented:
The one I have tested on has:
FC: MAMBA DJI F722 MK2
ESC: MAMBA F50_BL32

MartinWa commented:
Tested the RC6 and I see only below 1% errors (10th of a percent actually) so it looks like this fix did the trick
![image](https://user-images.githubusercontent.com/497015/165940145-3714a842-0b17-44b1-b854-1166421e1397.png)

Thanks to all Betaflight devs for all your work!

hydra commented:
> Tested the RC6 and I see only below 1% errors (10th of a percent actually) so it looks like this fix did the trick

@MartinWa it looks like you might still need updated ESC firmware as you should be getting 0% errors on the bench.  Did you try new BL Heli 'test code' that @sskaug posted?  If not please try that and report back again.
